### ISSUE TITLE
Intermittent failures when querying over TCP, possibly due to keep-alives

### ISSUE THREAD

**Issue by ivanr** [dnsjava/dnsjava#223]
We're using dnsjava (currently at 3.4.1) in our application to test configuration of DNS servers. We've noticed intermittent failures of TCP queries with servers that otherwise have no problem responding over TCP. Looking at the network capture, I notice that there appears to be a TCP keep-alive mechanism in dnsjava. In the one failure I managed to capture, the first query over TCP works fine, but the subsequent query fails and we observe a RST from the server.

How can we disable TCP keep alives? They are not appropriate for our use case because we'd like our zone tests to be deterministic. At the moment, our test of one zone is reusing a TCP connection of some other zone and we would like to avoid that.

**Comment by ibauersachs** [comment]
dnsjava does nothing special with the TCP socket it gets from Java. Isn't that just default TCP behavior as controlled by SO_KEEPALIVE via the operating system?

**Comment by ivanr** [comment]
Is dnsjava expected to be closing the TCP socket after a response is received? According to the network capture I am looking at it's not.

**Comment by ibauersachs** [comment]
The TCP is intended to be reused until it times out and is shared among the resolvers. The RST in the screenshot you sent me privately is coming from the server, as far as I can guess from the timestamp after about 1s of inactivity. It's been a while since I did the heavy lifting there, but I remember this as normal. DNS servers do some housekeeping, and you may be able to influence that with the EDNS timeout options.

**Comment by ivanr** [comment]
In the capture, a query (on a reused TCP connection) is sent but there is no response from the server, and 15 seconds later the connection is closed. This may or may not be a problem with this server.

Is there a way to configure dnsjava to always open a new TCP connection for every new query?

**Comment by ibauersachs** [comment]
> In the capture, a query (on a reused TCP connection) is sent but there is no response from the server, and 15 seconds later the connection is closed. This may or may not be a problem with this server.

I don't know why the server's not responding, maybe you're simply not allowed to perform queries? Is it under you control so that you could investigate the logs there?

> Is there a way to configure dnsjava to always open a new TCP connection for every new query?

No, there's not. And it would be horribly inefficient: that's what UDP is good for.

**Comment by ivanr** [comment]
The server is Cloudflare's. From what I can tell, then second query on a TCP connection always fails.

`dig +tcp +keepopen cloudflare.com soa @173.245.58.118 cloudflare.com soa @173.245.58.118`


> > Is there a way to configure dnsjava to always open a new TCP connection for every new query?
>
> No, there's not. And it would be horribly inefficient: that's what UDP is good for.

I'd prefer a simpler approach without a connection pool that's maybe less performant, but more reliable. As it is, we now have these very long timeouts and failures that we have to work around. Overall, we're less efficient than we'd be without a connection pool.

I figure that, for what we're doing, we now have dnsjava keeping a lot of connections we don't need open. From a quick look at the code, I didn't see where the connections are closed [by dnsjava]. If you're relying on the remote servers to close them, then perhaps there will be some servers that don't, keeping these connections open forever and wasting resources on both ends.

**Comment by ibauersachs** [comment]
> 
> 
> The server is Cloudflare's. From what I can tell, then second query on a TCP connection always fails. (EDIT I will report this to them.)
> 
> `dig +tcp +keepopen cloudflare.com soa @173.245.58.118 cloudflare.com soa @173.245.58.118`

I'd say the server is broken or intentionally doing some rather weird rate limiting because it's not meant to be used as a recursive resolver (see the dig output `;; WARNING: recursion requested but not available`). If you add `+keepalive` to indicate the EDNS keepalive option it outright refuses to answer at all. I don't why someone would do something like that. If a server operator doesn't want keep alive, just immediately send an RST after the answer instead of keeping it open.

The same command/query against 1.1.1.1 or 8.8.8.8 works just fine.

> > > Is there a way to configure dnsjava to always open a new TCP connection for every new query?
> > 
> > 
> > No, there's not. And it would be horribly inefficient: that's what UDP is good for.
> 
> I'd prefer a simpler approach without a connection pool that's maybe less performant, but more reliable. As it is, we now have these very long timeouts and failures that we have to work around. Overall, we're less efficient than we'd be without a connection pool.

Closing a TCP connection after each query would clearly go against what [RFC 7766, 6.2](https://datatracker.ietf.org/doc/html/rfc7766#section-6.2.1.1) says. I won't implement that.

I'm not sure the built-in resolvers of dnsjava are suited to your usecase.

> I figure that, for what we're doing, we now have dnsjava keeping a lot of connections we don't need open. From a quick look at the code, I didn't see where the connections are closed [by dnsjava]. If you're relying on the remote servers to close them, then perhaps there will be some servers that don't, keeping these connections open forever and wasting resources on both ends.

True, there needs to be a client-side timeout too which is currently missing, and specifically honoring the EDNS keep-alive in answers (if present).

**Comment by ivanr** [comment]
> I'd say the server is broken or intentionally doing some rather weird rate limiting because it's not meant to be used as a recursive resolver 

Indeed, it's an authoritative server and we're querying it as such.


> Closing a TCP connection after each query would clearly go against what RFC 7766, 6.2 says. 

The RFC uses SHOULD, which allows for a different approach where that makes sense, and it does in our case. Frankly, I don't think a library should carry strong opinions. The end user (application) is, by definition, in a better position to make decisions of what behaviour is appropriate.


> I won't implement that.

Understood. We'll work around it.
